File: test/language/parameterized_types/cpp98/SimpleParamTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "parameterized_types/simple_param/Item.h"

namespace parameterized_types
{
namespace simple_param
{

class ParameterizedTypesSimpleParamTest : public ::testing::Test
{
protected:
    void writeItemToByteArray(zserio::BitStreamWriter& writer, uint32_t version, uint16_t param,
            uint32_t extraParam)
    {
        writer.writeBits(param, 16);
        if (version >= HIGHER_VERSION)
            writer.writeBits(extraParam, 32);
    }

    void checkItemInBitStream(zserio::BitStreamReader& reader, uint32_t version, const Item& item)
    {
        ASSERT_EQ(item.getParam(), reader.readBits(16));
        if (version >= HIGHER_VERSION)
            ASSERT_EQ(item.getExtraParam(), reader.readBits(32));
    }

    static const uint32_t LOWER_VERSION;
    static const uint32_t HIGHER_VERSION;

    static const uint16_t ITEM_PARAM;
    static const uint32_t ITEM_EXTRA_PARAM;

    static const size_t ITEM_BIT_SIZE_WITHOUT_OPTIONAL;
    static const size_t ITEM_BIT_SIZE_WITH_OPTIONAL;
};

const uint32_t ParameterizedTypesSimpleParamTest::LOWER_VERSION = 9;
const uint32_t ParameterizedTypesSimpleParamTest::HIGHER_VERSION = 10;

const uint16_t ParameterizedTypesSimpleParamTest::ITEM_PARAM = 0xAA;
const uint32_t ParameterizedTypesSimpleParamTest::ITEM_EXTRA_PARAM = 0xBB;

const size_t ParameterizedTypesSimpleParamTest::ITEM_BIT_SIZE_WITHOUT_OPTIONAL = 16;
const size_t ParameterizedTypesSimpleParamTest::ITEM_BIT_SIZE_WITH_OPTIONAL = 16 + 32;

TEST_F(ParameterizedTypesSimpleParamTest, emptyConstructor)
{
    const uint16_t version = LOWER_VERSION;
    Item item;
    item.initialize(version);
    ASSERT_EQ(version, item.getVersion());
    ASSERT_FALSE(item.hasExtraParam());
}

TEST_F(ParameterizedTypesSimpleParamTest, bitStreamReaderConstructor)
{
    const uint32_t version = HIGHER_VERSION;
    zserio::BitStreamWriter writer;
    const uint16_t itemParam = ITEM_PARAM;
    const uint32_t itemExtraParam = ITEM_EXTRA_PARAM;
    writeItemToByteArray(writer, version, itemParam, itemExtraParam);

    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    Item item(reader, version);
    ASSERT_EQ(version, item.getVersion());
    ASSERT_EQ(itemParam, item.getParam());
    ASSERT_TRUE(item.hasExtraParam());
    ASSERT_EQ(itemExtraParam, item.getExtraParam());
}

TEST_F(ParameterizedTypesSimpleParamTest, copyConstructor)
{
    Item item;
    item.setParam(ITEM_PARAM);
    item.setExtraParam(ITEM_EXTRA_PARAM);

    const Item itemCopy1(item);
    ASSERT_THROW(itemCopy1.getVersion(), zserio::CppRuntimeException);

    item.initialize(HIGHER_VERSION);
    const Item itemCopy2(item);
    ASSERT_EQ(item.getVersion(), itemCopy2.getVersion());
}

TEST_F(ParameterizedTypesSimpleParamTest, operatorAssignment)
{
    Item item;
    item.setParam(ITEM_PARAM);
    item.setExtraParam(ITEM_EXTRA_PARAM);

    const Item itemCopy1 = item;
    ASSERT_THROW(itemCopy1.getVersion(), zserio::CppRuntimeException);

    item.initialize(HIGHER_VERSION);
    const Item itemCopy2 = item;
    ASSERT_EQ(item.getVersion(), itemCopy2.getVersion());
}

TEST_F(ParameterizedTypesSimpleParamTest, bitSizeOf)
{
    Item item1;
    item1.initialize(LOWER_VERSION);
    item1.setParam(ITEM_PARAM);
    const size_t expectedBitSizeWithoutOptional = ITEM_BIT_SIZE_WITHOUT_OPTIONAL;
    ASSERT_EQ(expectedBitSizeWithoutOptional, item1.bitSizeOf());

    Item item2;
    item2.initialize(HIGHER_VERSION);
    item2.setParam(ITEM_PARAM);
    item2.setExtraParam(ITEM_EXTRA_PARAM);
    const size_t expectedBitSizeWithOptional = ITEM_BIT_SIZE_WITH_OPTIONAL;
    ASSERT_EQ(expectedBitSizeWithOptional, item2.bitSizeOf());
}

TEST_F(ParameterizedTypesSimpleParamTest, initializeOffsets)
{
    Item item1;
    item1.initialize(LOWER_VERSION);
    item1.setParam(ITEM_PARAM);
    const size_t bitPosition = 1;
    const size_t expectedBitPositionWithoutOptional = bitPosition + ITEM_BIT_SIZE_WITHOUT_OPTIONAL;
    ASSERT_EQ(expectedBitPositionWithoutOptional, item1.initializeOffsets(bitPosition));

    Item item2;
    item2.initialize(HIGHER_VERSION);
    item2.setParam(ITEM_PARAM);
    item2.setExtraParam(ITEM_EXTRA_PARAM);
    const size_t expectedBitPositionWithOptional = bitPosition + ITEM_BIT_SIZE_WITH_OPTIONAL;
    ASSERT_EQ(expectedBitPositionWithOptional, item2.initializeOffsets(bitPosition));
}

TEST_F(ParameterizedTypesSimpleParamTest, operatorEquality)
{
    Item item1;
    item1.initialize(LOWER_VERSION);
    Item item2;
    item2.initialize(LOWER_VERSION);
    ASSERT_TRUE(item1 == item2);

    Item item3;
    item3.initialize(HIGHER_VERSION);

    ASSERT_FALSE(item2 == item3);
}

TEST_F(ParameterizedTypesSimpleParamTest, hashCode)
{
    Item item1;
    item1.initialize(LOWER_VERSION);
    Item item2;
    item2.initialize(LOWER_VERSION);
    ASSERT_EQ(item1.hashCode(), item2.hashCode());

    Item item3;
    item3.initialize(HIGHER_VERSION);
    ASSERT_TRUE(item2.hashCode() != item3.hashCode());
}

TEST_F(ParameterizedTypesSimpleParamTest, write)
{
    const uint32_t version = HIGHER_VERSION;
    Item item;
    item.initialize(version);
    item.setParam(ITEM_PARAM);
    item.setExtraParam(ITEM_EXTRA_PARAM);

    zserio::BitStreamWriter writer;
    item.write(writer);

    size_t writerBufferByteSize;
    const uint8_t* writerBuffer = writer.getWriteBuffer(writerBufferByteSize);
    zserio::BitStreamReader reader(writerBuffer, writerBufferByteSize);
    checkItemInBitStream(reader, version, item);
    reader.setBitPosition(0);

    const Item readItem(reader, version);
    ASSERT_EQ(item, readItem);
}

} // namespace compound_param
} // namespace parameterized_types

File: test/language/parameterized_types/java/parameterized_types/simple_param/SimpleParamTest.java
package parameterized_types.simple_param;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;

import javax.imageio.stream.FileImageInputStream;
import javax.imageio.stream.FileImageOutputStream;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.FileBitStreamReader;

public class SimpleParamTest
{
    @Test
    public void parameterConstructor()
    {
        final Item item = new Item(LOWER_VERSION);
        assertFalse(item.isExtraParamUsed());
    }

    @Test
    public void fileConstructor() throws IOException
    {
        final File file = new File("test.bin");
        final long version = LOWER_VERSION;
        writeItemToFile(file, version, ITEM_PARAM, ITEM_EXTRA_PARAM);
        final Item item = new Item(file, version);
        assertEquals(ITEM_PARAM, item.getParam());
        assertFalse(item.isExtraParamUsed());
    }

    @Test
    public void bitStreamReaderConstructor() throws IOException
    {
        final File file = new File("test.bin");
        final long version = HIGHER_VERSION;
        writeItemToFile(file, version, ITEM_PARAM, ITEM_EXTRA_PARAM);
        final BitStreamReader stream = new FileBitStreamReader(file);
        final Item item = new Item(stream, version);
        assertEquals(ITEM_PARAM, item.getParam());
        assertTrue(item.isExtraParamUsed());
        assertEquals(Long.valueOf(ITEM_EXTRA_PARAM), item.getExtraParam());
    }

    @Test
    public void fieldConstructor() throws IOException
    {
        final Item item = new Item(HIGHER_VERSION, ITEM_PARAM, (long)ITEM_EXTRA_PARAM);
        assertEquals(ITEM_PARAM, item.getParam());
        assertTrue(item.isExtraParamUsed());
        assertEquals(Long.valueOf(ITEM_EXTRA_PARAM), item.getExtraParam());
    }

    @Test
    public void bitSizeOf()
    {
        final Item item1 = new Item(LOWER_VERSION, ITEM_PARAM, Long.valueOf(ITEM_EXTRA_PARAM));
        assertEquals(ITEM_BIT_SIZE_WITHOUT_OPTIONAL, item1.bitSizeOf());

        final Item item2 = new Item(HIGHER_VERSION, ITEM_PARAM, Long.valueOf(ITEM_EXTRA_PARAM));
        assertEquals(ITEM_BIT_SIZE_WITH_OPTIONAL, item2.bitSizeOf());
    }

    @Test
    public void equals()
    {
        final Item item1 = new Item(LOWER_VERSION);
        final Item item2 = new Item(LOWER_VERSION);
        assertTrue(item1.equals(item2));

        final Item item3 = new Item(HIGHER_VERSION);
        assertFalse(item2.equals(item3));
    }

    @Test
    public void hashCodeMethod()
    {
        final Item item1 = new Item(LOWER_VERSION);
        final Item item2 = new Item(LOWER_VERSION);
        assertEquals(item1.hashCode(), item2.hashCode());

        final Item item3 = new Item(HIGHER_VERSION);
        assertTrue(item2.hashCode() != item3.hashCode());
    }

    @Test
    public void initializeOffsets()
    {
        final Item item1 = new Item(LOWER_VERSION, ITEM_PARAM, Long.valueOf(ITEM_EXTRA_PARAM));
        final int bitPosition = 1;
        assertEquals(bitPosition + ITEM_BIT_SIZE_WITHOUT_OPTIONAL, item1.initializeOffsets(bitPosition));

        final Item item2 = new Item(HIGHER_VERSION, ITEM_PARAM, Long.valueOf(ITEM_EXTRA_PARAM));
        assertEquals(bitPosition + ITEM_BIT_SIZE_WITH_OPTIONAL, item2.initializeOffsets(bitPosition));
    }

    @Test
    public void fileWrite() throws IOException
    {
        final long version1 = LOWER_VERSION;
        final Item item1 = new Item(version1);
        item1.setParam(ITEM_PARAM);
        final File file = new File("test.bin");
        item1.write(file);
        checkItemInFile(file, item1, version1);
        Item readItem1 = new Item(file, version1);
        assertEquals(item1, readItem1);

        final long version2 = HIGHER_VERSION;
        final Item item2 = new Item(version2, ITEM_PARAM, (long)ITEM_EXTRA_PARAM);
        item2.write(file);
        checkItemInFile(file, item2, version2);
        Item readItem2 = new Item(file, version2);
        assertEquals(item2, readItem2);
    }

    private void writeItemToFile(File file, long version, short param, int extraParam) throws IOException
    {
        final FileImageOutputStream stream = new FileImageOutputStream(file);
        stream.writeShort(param);
        if (version >= HIGHER_VERSION)
            stream.writeInt(extraParam);
        stream.close();
    }

    private void checkItemInFile(File file, Item item, long version) throws IOException
    {
        final FileImageInputStream stream = new FileImageInputStream(file);
        assertEquals(item.getParam(), stream.readUnsignedShort());
        if (version >= HIGHER_VERSION)
            assertEquals(item.getExtraParam(), Long.valueOf(stream.readUnsignedInt()));
        stream.close();
    }

    private static final long LOWER_VERSION = 9;
    private static final long HIGHER_VERSION = 10;

    private static final short ITEM_PARAM = 0xAA;
    private static final int ITEM_EXTRA_PARAM = 0xBB;

    private static final int ITEM_BIT_SIZE_WITHOUT_OPTIONAL = 16;
    private static final int ITEM_BIT_SIZE_WITH_OPTIONAL = 16 + 32;
}
